File: frontend/src/api/district-descriptions.ts
import api from "../utils/http";

export type DistrictDescription = {
  id: number;
  district_name: string;
  description: string | null;
  created_at: string;
  updated_at?: string | null;
};

export type DistrictDescriptionCreate = {
  district_name: string;
  description?: string | null;
};

export type DistrictDescriptionUpdate = {
  description?: string | null;
};

export const getDistrictDescriptions = async (): Promise<DistrictDescription[]> => {
  const { data } = await api.get<DistrictDescription[]>("/district-descriptions");
  return data;
};

export const getDistrictDescription = async (districtName: string): Promise<DistrictDescription | null> => {
  const { data } = await api.get<DistrictDescription | null>(`/district-descriptions/${encodeURIComponent(districtName)}`);
  return data;
};

export const createOrUpdateDistrictDescription = async (payload: DistrictDescriptionCreate): Promise<DistrictDescription> => {
  const { data } = await api.post<DistrictDescription>("/district-descriptions", payload);
  return data;
};

export const updateDistrictDescription = async (districtName: string, payload: DistrictDescriptionUpdate): Promise<DistrictDescription> => {
  const { data } = await api.put<DistrictDescription>(`/district-descriptions/${encodeURIComponent(districtName)}`, payload);
  return data;
};

export const deleteDistrictDescription = async (districtName: string): Promise<void> => {
  await api.delete(`/district-descriptions/${encodeURIComponent(districtName)}`);
};

File: backend/app/api/v1/routes/district_descriptions.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.district_description import DistrictDescription
from app.schemas.district_description import (
    DistrictDescription as DistrictDescriptionSchema,
    DistrictDescriptionCreate,
    DistrictDescriptionUpdate,
)
from app.core.security import require_roles

router = APIRouter()


@router.get("", response_model=List[DistrictDescriptionSchema])
def list_district_descriptions(
    db: Session = Depends(get_db),
    _: dict = Depends(require_roles("admin", "editor", "editor_plus")),
):
    """Получить все описания районов."""
    return db.query(DistrictDescription).all()


@router.get("/{district_name}", response_model=Optional[DistrictDescriptionSchema])
def get_district_description(
    district_name: str,
    db: Session = Depends(get_db),
    _: dict = Depends(require_roles("admin", "editor", "editor_plus")),
):
    """Получить описание района по имени."""
    desc = db.query(DistrictDescription).filter(
        DistrictDescription.district_name == district_name
    ).first()
    return desc


@router.post("", response_model=DistrictDescriptionSchema)
def create_or_update_district_description(
    payload: DistrictDescriptionCreate,
    db: Session = Depends(get_db),
    _: dict = Depends(require_roles("admin", "editor", "editor_plus")),
):
    """Создать или обновить описание района."""
    existing = db.query(DistrictDescription).filter(
        DistrictDescription.district_name == payload.district_name
    ).first()
    
    if existing:
        existing.description = payload.description
        db.commit()
        db.refresh(existing)
        return existing
    else:
        new_desc = DistrictDescription(
            district_name=payload.district_name,
            description=payload.description,
        )
        db.add(new_desc)
        db.commit()
        db.refresh(new_desc)
        return new_desc


@router.put("/{district_name}", response_model=DistrictDescriptionSchema)
def update_district_description(
    district_name: str,
    payload: DistrictDescriptionUpdate,
    db: Session = Depends(get_db),
    _: dict = Depends(require_roles("admin", "editor", "editor_plus")),
):
    """Обновить описание района."""
    desc = db.query(DistrictDescription).filter(
        DistrictDescription.district_name == district_name
    ).first()
    
    if not desc:
        # Создаем новую запись
        desc = DistrictDescription(
            district_name=district_name,
            description=payload.description,
        )
        db.add(desc)
    else:
        desc.description = payload.description
    
    db.commit()
    db.refresh(desc)
    return desc


@router.delete("/{district_name}")
def delete_district_description(
    district_name: str,
    db: Session = Depends(get_db),
    _: dict = Depends(require_roles("admin")),
):
    """Удалить описание района."""
    desc = db.query(DistrictDescription).filter(
        DistrictDescription.district_name == district_name
    ).first()
    
    if desc:
        db.delete(desc)
        db.commit()
    
    return {"ok": True}
